src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -249,16 +246,53 @@ public void leaveToken(DetailAST ast) {
     /**
      * Checks whether a class should be declared as final or not.
      *
-     * @param desc description of the class
+     * @param classDesc description of the class
      * @return true if given class should be declared as final otherwise false
      */
-    private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
-        return desc.isWithPrivateCtor()
-                && !(desc.isDeclaredAsAbstract()
-                    || desc.isSuperClassOfAnonymousInnerClass())
-                && !desc.isDeclaredAsFinal()
-                && !desc.isWithNonPrivateCtor()
-                && !desc.isWithNestedSubclass();
+    private static boolean shouldBeDeclaredAsFinal(ClassDesc classDesc) {
+        final boolean isAbstract = classDesc.isDeclaredAsAbstract();
+        final boolean isSuperClassOfAnonymousInnerClass =
+                classDesc.isSuperClassOfAnonymousInnerClass();
+        final boolean declaredAsFinal = classDesc.isDeclaredAsFinal();
+        final boolean withNestedSubClass = classDesc.isWithNestedSubclass();
+        final boolean shouldBeFinal;
+
+        if (classDesc.isWithNonPrivateCtor()) {
+            shouldBeFinal = false;
+        }
+        else if (hasImplicitConstructor(classDesc.getTypeDeclarationAst())) {
+            shouldBeFinal = !(isAbstract || isSuperClassOfAnonymousInnerClass)
+                    && !declaredAsFinal
+                    && !withNestedSubClass
+                    && isClassPrivate(classDesc.getTypeDeclarationAst());
+        }
+        else {
+            shouldBeFinal = !(isAbstract || isSuperClassOfAnonymousInnerClass)
+                    && !declaredAsFinal
+                    && !withNestedSubClass;
+        }
+        return shouldBeFinal;
+    }
+
+    /**
+     * Check it the constructor is default. default means there is no constructor
+     * defined.
+     *
+     * @param desc class to check
+     * @return boolean value for default constructor
+     */
+    private static boolean hasImplicitConstructor(DetailAST desc) {
+        return desc.getLastChild().findFirstToken(TokenTypes.CTOR_DEF) == null;
```

Comment:
move this method to ClassDesc, and evaluate it it ones to return pre-calculted result always.
Example of already similar methods: `classDesc.isDeclaredAsAbstract();`

Reply:
done